### PR TITLE
[CIFuzz] Check fuzzers

### PR DESCRIPTION
**Pull request by Leo-Neat**
Checks that the fuzzer build was a success in the build action.

### CODE REVIEW

File: infra/cifuzz/cifuzz.py [hunk 1]
```diff
@@ -151,7 +151,7 @@ def build_fuzzers(project_name,
   if helper.docker_run(command):
     logging.error('Building fuzzers failed.')
     return False
-  return True
+  return not helper.check_build_impl(project_name, None)
```
Comment (jonathanmetzman) [inline infra/cifuzz/cifuzz.py]: Can we call this from the caller of build_fuzzers?

Comment (jonathanmetzman) [reply infra/cifuzz/cifuzz.py]: Want to keep the code modular.

File: infra/cifuzz/actions/build_fuzzers/build_fuzzers_entrypoint.py [hunk 2]
```diff
@@ -78,7 +81,7 @@ def main():
     logging.error('Error building fuzzers for project %s with pull request %s.',
                   oss_fuzz_project_name, pr_ref)
     return error_code
-  return 0
+  return not helper.check_build_impl(project_name, None)
```
Comment (jonathanmetzman) [inline infra/cifuzz/actions/build_fuzzers/build_fuzzers_entrypoint.py]: Isn't this supposed to be returning 0 on success, I think this will return True.

Comment (Leo-Neat) [reply infra/cifuzz/actions/build_fuzzers/build_fuzzers_entrypoint.py]: Fixed.

File: infra/cifuzz/cifuzz.py (hunk 1 above)
Comment (Leo-Neat) [reply infra/cifuzz/cifuzz.py]: Done.

**Comment by evverx** [comment]
@Leo-Neat thank you!

I haven't figured out how to pass `ALLOWED_BROKEN_TARGETS_PERCENTAGE=0` to `test_all`. Given that it's kind of important to `systemd` at least: https://github.com/google/oss-fuzz/pull/2419, I'm wondering whether it's going to be addressed later?

**Comment by jonathanmetzman** [comment]
> @Leo-Neat thank you!
> 
> I haven't figured out how to pass `ALLOWED_BROKEN_TARGETS_PERCENTAGE=0` to `test_all`. Given that it's kind of important to `systemd` at least: #2419, I'm wondering whether it's going to be addressed later?

Is this being used for official OSS-Fuzz builds of systemd? I assume so, but don't know how.

File: infra/cifuzz/cifuzz.py [hunk 3]
```diff
@@ -203,6 +209,40 @@ def run_fuzzers(fuzz_seconds, workspace, project_name):
   return True, False
 
 
+def check_fuzzer_build(out_dir):
+  """Checks the integrity of the built fuzzers.
+
+  Args:
+    out_dir: The location of the fuzzer binaries.
```
Comment (jonathanmetzman) [inline infra/cifuzz/cifuzz.py]: nit: "The directory containing the fuzzer binaries"

File: infra/cifuzz/cifuzz.py [hunk 4]
```diff
@@ -203,6 +209,40 @@ def run_fuzzers(fuzz_seconds, workspace, project_name):
...
+
+  Returns:
+    True if fuzzers are correct or False if not.
```
Comment (jonathanmetzman) [inline infra/cifuzz/cifuzz.py]: I think the style I usually see is to leave the "False if not" part implicit.

File: infra/cifuzz/cifuzz_test.py [hunk 5]
```diff
@@ -217,5 +217,22 @@ def test_parse_invalid_output(self):
...
+    """Checks that the function returns true for valid fuzzer."""
+    test_fuzzer_dir = os.path.join(TEST_FILES_PATH, 'out')
+    self.assertTrue(cifuzz.check_fuzzer_build(test_fuzzer_dir))
+
+  def test_not_a_valid_fuzz_path(self):
+    """Tests that false is returned when a bad path is given."""
+    self.assertFalse(cifuzz.check_fuzzer_build('not/a/valid/path'))
+
+  def test_not_a_valid_fuzzer(self):
+    """Checks a directory that exists but does not have fuzzzers is False."""
```
Comment (jonathanmetzman) [inline infra/cifuzz/cifuzz_test.py]: "fuzzers"

File: infra/cifuzz/cifuzz_test.py [hunk 6]
```diff
@@ -217,5 +217,22 @@ def test_parse_invalid_output(self):
...
+class CheckFuzzerBuildUnitTest(unittest.TestCase):
+  """Tests the check_fuzzer_build function in the cifuzz module."""
+
+  def test_correct_fuzzer_build(self):
+    """Checks that the function returns true for valid fuzzer."""
+    test_fuzzer_dir = os.path.join(TEST_FILES_PATH, 'out')
+    self.assertTrue(cifuzz.check_fuzzer_build(test_fuzzer_dir))
+
+  def test_not_a_valid_fuzz_path(self):
+    """Tests that false is returned when a bad path is given."""
```
Comment (jonathanmetzman) [inline infra/cifuzz/cifuzz_test.py]: "False"

File: infra/cifuzz/cifuzz_test.py [hunk 7]
```diff
@@ -217,5 +217,22 @@ def test_parse_invalid_output(self):
       self.assertEqual(len(os.listdir(tmp_dir)), 0)
 
 
+class CheckFuzzerBuildUnitTest(unittest.TestCase):
+  """Tests the check_fuzzer_build function in the cifuzz module."""
+
+  def test_correct_fuzzer_build(self):
+    """Checks that the function returns true for valid fuzzer."""
```
Comment (jonathanmetzman) [inline infra/cifuzz/cifuzz_test.py]: "Checks that check_fuzzer_build function returns True for valid fuzzers"

File: infra/cifuzz/cifuzz.py [hunk 8]
```diff
@@ -203,6 +209,40 @@ def run_fuzzers(fuzz_seconds, workspace, project_name):
...
+  """
+  if not os.path.exists(out_dir):
+    logging.error('Invalid out directory: %s.', out_dir)
+    return False
+  if not os.listdir(out_dir):
+    logging.error('No fuzzers found in out directory: %s.', out_dir)
+    return False
+
+  command = [
+      '--cap-add', 'SYS_PTRACE', '-e', 'FUZZING_ENGINE=' + DEFAULT_ENGINE, '-e',
```
Comment (jonathanmetzman) [inline infra/cifuzz/cifuzz.py]: if formatting allows it, I'd put each `'-e'` on the same line as the env variable it is setting.

File: infra/cifuzz/cifuzz.py (hunk 3 above)
Comment (Leo-Neat) [reply infra/cifuzz/cifuzz.py]: Fixed.

File: infra/cifuzz/cifuzz.py (hunk 4 above)
Comment (Leo-Neat) [reply infra/cifuzz/cifuzz.py]: Removed.

File: infra/cifuzz/cifuzz.py (hunk 8 above)
Comment (Leo-Neat) [reply infra/cifuzz/cifuzz.py]: The formatter is not happy with the change.

File: infra/cifuzz/cifuzz_test.py (hunk 5 above)
Comment (Leo-Neat) [reply infra/cifuzz/cifuzz_test.py]: Fixed.

File: infra/cifuzz/cifuzz_test.py (hunk 7 above)
Comment (Leo-Neat) [reply infra/cifuzz/cifuzz_test.py]: Fixed.

File: infra/cifuzz/cifuzz_test.py (hunk 6 above)
Comment (Leo-Neat) [reply infra/cifuzz/cifuzz_test.py]: Fixed.

**Comment by evverx** [comment]
> Is this being used for official OSS-Fuzz builds of systemd?

It's used in the sense that when PRs are opened on GitHub, `helper.py check_build -e ALLOWED_BROKEN_TARGETS_PERCENTAGE=0` is automatically run by a bash script to make sure the fuzz targets are more or less fine. I hope the bash script can be eventually replaced by CIFuzz.

**Comment by jonathanmetzman** [comment]
> > Is this being used for official OSS-Fuzz builds of systemd?
> 
> It's used in the sense that when PRs are opened on GitHub, `helper.py check_build -e ALLOWED_BROKEN_TARGETS_PERCENTAGE=0` is automatically run by a bash script to make sure the fuzz targets are more or less fine. I hope the bash script can be eventually replaced by CIFuzz.

Ah got it.

**Comment by Leo-Neat** [comment]
> > @Leo-Neat thank you!
> > I haven't figured out how to pass `ALLOWED_BROKEN_TARGETS_PERCENTAGE=0` to `test_all`. Given that it's kind of important to `systemd` at least: #2419, I'm wondering whether it's going to be addressed later?
> 
> Is this being used for official OSS-Fuzz builds of systemd? I assume so, but don't know how.

@evverx You should be able to just set `ALLOWED_BROKEN_TARGETS_PERCENTAGE` in the docker file before calling  `test_all`

**Comment by evverx** [comment]
@Leo-Neat I'd prefer not to touch the docker file. In fact, after running into https://github.com/systemd/systemd/pull/14520 I was going to trim it by moving almost everything from there to our repository.

**Comment by Leo-Neat** [comment]
@evverx  Sorry I misspoke. I meant set the environment variable in the GitHub action's config file under the build fuzzers action.

**Comment by evverx** [comment]
@Leo-Neat as far as I know, `docker-run` doesn't pass environment variables from the environment where it's run to the processes inside containers. For it to work `check_fuzzer_build` would have to add something like `-e ALLOWED_BROKEN_TARGETS_PERCENTAGE=$ALLOWED_BROKEN_TARGETS_PERCENTAGE` to the command line.

File: infra/cifuzz/actions/build_fuzzers/build_fuzzers_entrypoint.py [hunk 9]
```diff
@@ -78,7 +78,10 @@ def main():
     logging.error('Error building fuzzers for project %s with pull request %s.',
                   oss_fuzz_project_name, pr_ref)
     return error_code
-  return 0
+  out_dir = os.path.join(workspace, 'out')
+  if cifuzz.check_fuzzer_build(out_dir):
```
Comment (Dor1s) [inline infra/cifuzz/actions/build_fuzzers/build_fuzzers_entrypoint.py]: should `dry_run` affect the check build behavior?

**Pull request merged**